File: joystickView/src/main/java/com/zerokol/views/joystickView/JoystickView.java
package com.zerokol.views.joystickView;

import android.content.Context;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.Paint;
import android.util.AttributeSet;
import android.view.MotionEvent;
import android.view.View;

public class JoystickView extends View implements Runnable {
    // Constants
    private final double RAD = 57.2957795;
    public final static long DEFAULT_LOOP_INTERVAL = 100; // 100 ms
    // That is the correct order 
    public final static int FRONT = 3;
    public final static int FRONT_RIGHT = 2;
    public final static int RIGHT = 1;
    public final static int RIGHT_BOTTOM = 8;
    public final static int BOTTOM = 7;
    public final static int BOTTOM_LEFT = 6;
    public final static int LEFT = 5;
    public final static int LEFT_FRONT = 4;
    // Variables
    private OnJoystickMoveListener onJoystickMoveListener; // Listener
    private Thread thread = new Thread(this);
    private long loopInterval = DEFAULT_LOOP_INTERVAL;
    private int xPosition = 0; // Touch x position
    private int yPosition = 0; // Touch y position
    private double centerX = 0; // Center view x position
    private double centerY = 0; // Center view y position
    private Paint mainCircle;
    private Paint secondaryCircle;
    private Paint button;
    private Paint horizontalLine;
    private Paint verticalLine;
    private int joystickRadius;
    private int buttonRadius;
    private int lastAngle = 0;
    private int lastPower = 0;

    public JoystickView(Context context) {
        super(context);
    }

    public JoystickView(Context context, AttributeSet attrs) {
        super(context, attrs);
        initJoystickView();
    }

    public JoystickView(Context context, AttributeSet attrs, int defaultStyle) {
        super(context, attrs, defaultStyle);
        initJoystickView();
    }

    protected void initJoystickView() {
        mainCircle = new Paint(Paint.ANTI_ALIAS_FLAG);
        mainCircle.setColor(Color.WHITE);
        mainCircle.setStyle(Paint.Style.FILL_AND_STROKE);

        secondaryCircle = new Paint();
        secondaryCircle.setColor(Color.GREEN);
        secondaryCircle.setStyle(Paint.Style.STROKE);

        verticalLine = new Paint();
        verticalLine.setStrokeWidth(5);
        verticalLine.setColor(Color.RED);

        horizontalLine = new Paint();
        horizontalLine.setStrokeWidth(2);
        horizontalLine.setColor(Color.BLACK);

        button = new Paint(Paint.ANTI_ALIAS_FLAG);
        button.setColor(Color.RED);
        button.setStyle(Paint.Style.FILL);
    }

    @Override
    protected void onFinishInflate() {
    }

    @Override
    protected void onSizeChanged(int xNew, int yNew, int xOld, int yOld) {
        super.onSizeChanged(xNew, yNew, xOld, yOld);
        // before measure, get the center of view
        xPosition = (int) getWidth() / 2;
        yPosition = (int) getWidth() / 2;
        int d = Math.min(xNew, yNew);
        buttonRadius = (int) (d / 2 * 0.25);
        joystickRadius = (int) (d / 2 * 0.75);

    }

    @Override
    protected void onMeasure(int widthMeasureSpec, int heightMeasureSpec) {
        // setting the measured values to resize the view to a certain width and
        // height
        int d = Math.min(measure(widthMeasureSpec), measure(heightMeasureSpec));

        setMeasuredDimension(d, d);

    }

    private int measure(int measureSpec) {
        int result = 0;

        // Decode the measurement specifications.
        int specMode = MeasureSpec.getMode(measureSpec);
        int specSize = MeasureSpec.getSize(measureSpec);

        if (specMode == MeasureSpec.UNSPECIFIED) {
            // Return a default size of 200 if no bounds are specified.
            result = 200;
        } else {
            // As you want to fill the available space
            // always return the full available bounds.
            result = specSize;
        }
        return result;
    }

    @Override
    protected void onDraw(Canvas canvas) {
        // super.onDraw(canvas);
        centerX = (getWidth()) / 2;
        centerY = (getHeight()) / 2;

        // painting the main circle
        canvas.drawCircle((int) centerX, (int) centerY, joystickRadius,
                mainCircle);
        // painting the secondary circle
        canvas.drawCircle((int) centerX, (int) centerY, joystickRadius / 2,
                secondaryCircle);
        // paint lines
        canvas.drawLine((float) centerX, (float) centerY, (float) centerX,
                (float) (centerY - joystickRadius), verticalLine);
        canvas.drawLine((float) (centerX - joystickRadius), (float) centerY,
                (float) (centerX + joystickRadius), (float) centerY,
                horizontalLine);
        canvas.drawLine((float) centerX, (float) (centerY + joystickRadius),
                (float) centerX, (float) centerY, horizontalLine);

        // painting the move button
        canvas.drawCircle(xPosition, yPosition, buttonRadius, button);
    }

    @Override
    public boolean onTouchEvent(MotionEvent event) {
        xPosition = (int) event.getX();
        yPosition = (int) event.getY();
        double abs = Math.sqrt((xPosition - centerX) * (xPosition - centerX)
                + (yPosition - centerY) * (yPosition - centerY));
        if (abs > joystickRadius) {
            xPosition = (int) ((xPosition - centerX) * joystickRadius / abs + centerX);
            yPosition = (int) ((yPosition - centerY) * joystickRadius / abs + centerY);
        }
        invalidate();
        if (event.getAction() == MotionEvent.ACTION_UP) {
            xPosition = (int) centerX;
            yPosition = (int) centerY;
            thread.interrupt();
            if (onJoystickMoveListener != null)
                onJoystickMoveListener.onValueChanged(getAngle(), getPower(),
                        getDirection());
        }
        if (onJoystickMoveListener != null
                && event.getAction() == MotionEvent.ACTION_DOWN) {
            if (thread != null && thread.isAlive()) {
                thread.interrupt();
            }
            thread = new Thread(this);
            thread.start();
            if (onJoystickMoveListener != null)
                onJoystickMoveListener.onValueChanged(getAngle(), getPower(),
                        getDirection());
        }
        return true;
    }

    private int getAngle() {
        if (xPosition > centerX) {
            if (yPosition < centerY) {
                return lastAngle = (int) (Math.atan((yPosition - centerY)
                        / (xPosition - centerX))
                        * RAD + 90);
            } else if (yPosition > centerY) {
                return lastAngle = (int) (Math.atan((yPosition - centerY)
                        / (xPosition - centerX)) * RAD) + 90;
            } else {
                return lastAngle = 90;
            }
        } else if (xPosition < centerX) {
            if (yPosition < centerY) {
                return lastAngle = (int) (Math.atan((yPosition - centerY)
                        / (xPosition - centerX))
                        * RAD - 90);
            } else if (yPosition > centerY) {
                return lastAngle = (int) (Math.atan((yPosition - centerY)
                        / (xPosition - centerX)) * RAD) - 90;
            } else {
                return lastAngle = -90;
            }
        } else {
            if (yPosition <= centerY) {
                return lastAngle = 0;
            } else {
                if (lastAngle < 0) {
                    return lastAngle = -180;
                } else {
                    return lastAngle = 180;
                }
            }
        }
    }

    private int getPower() {
        
        lastPower=(int) (100 * Math.sqrt((xPosition - centerX)
                * (xPosition - centerX) + (yPosition - centerY)
                * (yPosition - centerY)) / joystickRadius);

        return (int)(lastPower);
    }

    private int getDirection() {
        if (lastPower == 0 && lastAngle == 0) {
            return 0;
        }
        int a = 0;
        if (lastAngle <= 0) {
            a = (lastAngle * -1) + 90;
        } else if (lastAngle > 0) {
            if (lastAngle <= 90) {
                a = 90 - lastAngle;
            } else {
                a = 360 - (lastAngle - 90);
            }
        }

        int direction = (int) (((a + 22) / 45) + 1);

        if (direction > 8) {
            direction = 1;
        }
        return direction;
    }

    public void setOnJoystickMoveListener(OnJoystickMoveListener listener,
                                          long repeatInterval) {
        this.onJoystickMoveListener = listener;
        this.loopInterval = repeatInterval;
    }

    public interface OnJoystickMoveListener {
        public void onValueChanged(int angle, int power, int direction);
    }

    @Override
    public void run() {
        while (!Thread.interrupted()) {
            post(new Runnable() {
                public void run() {
                    if (onJoystickMoveListener != null)
                        onJoystickMoveListener.onValueChanged(getAngle(),
                                getPower(), getDirection());
                }
            });
            try {
                Thread.sleep(loopInterval);
            } catch (InterruptedException e) {
                break;
            }
        }
    }
}

File: joystickView/src/main/java/com/zerokol/views/joystickView/JoystickView.kt
package com.zerokol.views.joystickView

import android.content.Context
import android.graphics.Canvas
import android.graphics.Color
import android.graphics.Paint
import android.util.AttributeSet
import android.view.MotionEvent
import android.view.View

class JoystickView : View, Runnable {
    // Constants
    private val RAD = 57.2957795
    // Variables
    private var onJoystickMoveListener // Listener
            : OnJoystickMoveListener? = null
    private var thread: Thread? = Thread(this)
    private var loopInterval = DEFAULT_LOOP_INTERVAL
    private var xPosition = 0 // Touch x position
    private var yPosition = 0 // Touch y position
    private var centerX = 0.0 // Center view x position
    private var centerY = 0.0 // Center view y position
    private val mainCircle: Paint
    private val secondaryCircle: Paint
    private val button: Paint
    private val horizontalLine: Paint
    private val verticalLine: Paint
    private var joystickRadius = 0
    private var buttonRadius = 0
    private var lastAngle = 0
    private val lastPower = 0

    constructor(context: Context?) : super(context)
    constructor(context: Context?, attrs: AttributeSet?) : super(context, attrs)
    constructor(context: Context?, attrs: AttributeSet?, defaultStyle: Int) : super(context, attrs, defaultStyle)

    init {
        mainCircle = Paint(Paint.ANTI_ALIAS_FLAG)
        mainCircle.color = Color.WHITE
        mainCircle.style = Paint.Style.FILL_AND_STROKE
        secondaryCircle = Paint()
        secondaryCircle.color = Color.GREEN
        secondaryCircle.style = Paint.Style.STROKE
        verticalLine = Paint()
        verticalLine.strokeWidth = 5f
        verticalLine.color = Color.RED
        horizontalLine = Paint()
        horizontalLine.strokeWidth = 2f
        horizontalLine.color = Color.BLACK
        button = Paint(Paint.ANTI_ALIAS_FLAG)
        button.color = Color.RED
        button.style = Paint.Style.FILL
    }

    override fun onFinishInflate() {
        super.onFinishInflate()
    }
    override fun onSizeChanged(xNew: Int, yNew: Int, xOld: Int, yOld: Int) {
        super.onSizeChanged(xNew, yNew, xOld, yOld)
        // before measure, get the center of view
        xPosition = width / 2
        yPosition = width / 2
        val d = Math.min(xNew, yNew)
        buttonRadius = (d / 2 * 0.25).toInt()
        joystickRadius = (d / 2 * 0.75).toInt()
    }

    override fun onMeasure(widthMeasureSpec: Int, heightMeasureSpec: Int) { // setting the measured values to resize the view to a certain width and
// height
        val d = Math.min(measure(widthMeasureSpec), measure(heightMeasureSpec))
        setMeasuredDimension(d, d)
    }

    private fun measure(measureSpec: Int): Int {
        var result = 0
        // Decode the measurement specifications.
        val specMode = MeasureSpec.getMode(measureSpec)
        val specSize = MeasureSpec.getSize(measureSpec)
        result = if (specMode == MeasureSpec.UNSPECIFIED) { // Return a default size of 200 if no bounds are specified.
            200
        } else { // As you want to fill the available space
// always return the full available bounds.
            specSize
        }
        return result
    }

    override fun onDraw(canvas: Canvas) { // super.onDraw(canvas);
        centerX = width / 2.toDouble()
        centerY = height / 2.toDouble()
        // painting the main circle
        canvas.drawCircle(centerX.toFloat(), centerY.toFloat(), joystickRadius.toFloat(),
                mainCircle)
        // painting the secondary circle
        canvas.drawCircle(centerX.toFloat(), centerY.toFloat(), joystickRadius / 2.toFloat(),
                secondaryCircle)
        // paint lines
        canvas.drawLine(centerX.toFloat(), centerY.toFloat(), centerX.toFloat(),
                (centerY - joystickRadius).toFloat(), verticalLine)
        canvas.drawLine((centerX - joystickRadius).toFloat(), centerY.toFloat(),
                (centerX + joystickRadius).toFloat(), centerY.toFloat(),
                horizontalLine)
        canvas.drawLine(centerX.toFloat(), (centerY + joystickRadius).toFloat(),
                centerX.toFloat(), centerY.toFloat(), horizontalLine)
        // painting the move button
        canvas.drawCircle(xPosition.toFloat(), yPosition.toFloat(), buttonRadius.toFloat(), button)
    }

    override fun onTouchEvent(event: MotionEvent): Boolean {
        xPosition = event.x.toInt()
        yPosition = event.y.toInt()
        val abs = Math.sqrt((xPosition - centerX) * (xPosition - centerX)
                + (yPosition - centerY) * (yPosition - centerY))
        if (abs > joystickRadius) {
            xPosition = ((xPosition - centerX) * joystickRadius / abs + centerX).toInt()
            yPosition = ((yPosition - centerY) * joystickRadius / abs + centerY).toInt()
        }
        invalidate()
        if (event.action == MotionEvent.ACTION_UP) {
            xPosition = centerX.toInt()
            yPosition = centerY.toInt()
            thread!!.interrupt()
            if (onJoystickMoveListener != null) onJoystickMoveListener!!.onValueChanged(angle, power,
                    direction)
        }
        if (onJoystickMoveListener != null
                && event.action == MotionEvent.ACTION_DOWN) {
            if (thread != null && thread!!.isAlive) {
                thread!!.interrupt()
            }
            thread = Thread(this)
            thread!!.start()
            if (onJoystickMoveListener != null) onJoystickMoveListener!!.onValueChanged(angle, power,
                    direction)
        }
        return true
    }

    private val angle: Int
        private get() {
            if (xPosition > centerX) {
                if (yPosition < centerY) {
                    return (Math.atan((yPosition - centerY)
                            / (xPosition - centerX))
                            * RAD + 90).toInt().also{ lastAngle = it }
                } else return if (yPosition > centerY) {
                    (Math.atan((yPosition - centerY)
                            / (xPosition - centerX)) * RAD).toInt() + 90.also { lastAngle = it }
                } else {
                    90.also { lastAngle = it }
                }
            } else if (xPosition < centerX) {
                if (yPosition < centerY) {
                    return (Math.atan((yPosition - centerY)
                            / (xPosition - centerX))
                            * RAD - 90).toInt().also{ lastAngle = it }
                } else return if (yPosition > centerY) {
                    (Math.atan((yPosition - centerY)
                            / (xPosition - centerX)) * RAD).toInt() - 90.also { lastAngle = it }
                } else {
                    -90.also { lastAngle = it }
                }
            } else {
                return if (yPosition <= centerY) {
                    0.also { lastAngle = it }
                } else {
                    if (lastAngle < 0) {
                        -180.also { lastAngle = it }
                    } else {
                        180.also { lastAngle = it }
                    }
                }
            }
        }

    private val power: Int
        private get() = (100 * Math.sqrt((xPosition - centerX)
                * (xPosition - centerX) + (yPosition - centerY)
                * (yPosition - centerY)) / joystickRadius).toInt()

    private val direction: Int
        private get() {
            if (lastPower == 0 && lastAngle == 0) {
                return 0
            }
            var a = 0
            if (lastAngle <= 0) {
                a = lastAngle * -1 + 90
            } else if (lastAngle > 0) {
                a = if (lastAngle <= 90) {
                    90 - lastAngle
                } else {
                    360 - (lastAngle - 90)
                }
            }
            var direction = ((a + 22) / 45 + 1)
            if (direction > 8) {
                direction = 1
            }
            return direction
        }

    fun setOnJoystickMoveListener(listener: OnJoystickMoveListener?,
                                  repeatInterval: Long) {
        onJoystickMoveListener = listener
        loopInterval = repeatInterval
    }

    interface OnJoystickMoveListener {
        fun onValueChanged(angle: Int, power: Int, direction: Int)
    }

    override fun run() {
        while (!Thread.interrupted()) {
            post {
                if (onJoystickMoveListener != null) onJoystickMoveListener!!.onValueChanged(angle,
                        power, direction)
            }
            try {
                Thread.sleep(loopInterval)
            } catch (e: InterruptedException) {
                break
            }
        }
    }

    companion object {
        const val DEFAULT_LOOP_INTERVAL: Long = 100 // 100 ms
        const val FRONT = 3
        const val FRONT_RIGHT = 4
        const val RIGHT = 5
        const val RIGHT_BOTTOM = 6
        const val BOTTOM = 7
        const val BOTTOM_LEFT = 8
        const val LEFT = 1
        const val LEFT_FRONT = 2
    }
}